File: Lab4/Ex1/src/Buffer.java
import java.util.Arrays;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Buffer {
    public final int length;
    private final int[] data;
    private final Lock lock = new ReentrantLock();

    public Buffer(int length, int start) {
        this.length = length;
        this.data = new int[length];
        fillWIth(start);
    }

    private void fillWIth(int start) {
        for (int i = 0; i < length; i++)
            data[i] = start;
    }

    public Condition newCondition() {
        return lock.newCondition();
    }

    public void lock() {
        lock.lock();
    }

    public void unlock() {
        lock.unlock();
    }

    public void put(int index, int object) {
        data[index] = object;
    }

    public int get(int index) {
        return data[index];
    }

    public void log() {
        System.out.println("Buffer: " + this);
    }

    @Override
    public String toString() {
        return Arrays.toString(data);
    }
}

File: Lab4/Ex2/src/main/kotlin/Buffer.kt
import java.lang.RuntimeException
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.ReentrantLock

data class Buffer(val size: Int) {
    private val lock = ReentrantLock()
    private var state = 0

    val availableItems
        get() = state

    val availableSpace
        get() = size - state

    val producers: Condition = lock.newCondition()
    val consumers: Condition = lock.newCondition()

    fun put(amount: Int) {
        if (amount > availableSpace)
            throw RuntimeException("Not enough space in buffer - available $availableSpace, expected $amount")

        state += amount
    }

    fun get(amount: Int) {
        if (amount > state)
            throw RuntimeException("Not enough items in buffer - available $state, expected $amount")

        state -= amount
    }

    fun lock() {
        lock.lock()
    }

    fun unlock() {
        lock.unlock()
    }

    fun log() {
        println("Buffer: $state / $size")
    }
}